linux/ibus-keyman/src/engine.c

```diff
@@ -319,6 +319,115 @@ ibus_keyman_engine_init(IBusKeymanEngine *keyman) {
+      g_strfreev(taglang);
+    }
+    else {
+      lang = g_strdup(splitlang[0]);
+    }
+    g_strfreev(splitlang);
+  }
+  else {
+    lang = strdup("en-US");
+  }
+  g_message("lang is %s", lang);
+  km_kbp_cp *cp = g_utf8_to_utf16(lang, -1, NULL, NULL, NULL);
+  return cp;
+  // g_free(lang);
+#endif
+}
+
+static km_kbp_status
+setup_environment(IBusKeymanEngine *keyman)
+{
+  g_assert(keyman);
+  g_message("%s: setting up environment", __FUNCTION__);
+
+  // Allocate enough options for: 3 environments plus 1 pad struct of 0's
+  km_kbp_option_item *environment_opts = g_new0(km_kbp_option_item, 4);
+
+  environment_opts[0].scope = KM_KBP_OPT_ENVIRONMENT;
+  environment_opts[0].key   = KM_KBP_KMX_ENV_PLATFORM;
+  environment_opts[0].value = u"linux desktop hardware native";
+
+  environment_opts[1].scope = KM_KBP_OPT_ENVIRONMENT;
+  environment_opts[1].key   = KM_KBP_KMX_ENV_BASELAYOUT;
+  environment_opts[1].value = u"kbdus.dll";
+
+  environment_opts[2].scope = KM_KBP_OPT_ENVIRONMENT;
+  environment_opts[2].key   = KM_KBP_KMX_ENV_BASELAYOUTALT;
+  environment_opts[2].value = get_base_layout();  // TODO: free when mnemonic layouts are to be supported
+
+  environment_opts[3].scope = 0;
+  environment_opts[3].key   = 0;
+  environment_opts[3].value = NULL;
+
+  km_kbp_status status = km_kbp_state_create(keyman->keyboard, environment_opts, &(keyman->state));
+  if (status != KM_KBP_STATUS_OK) {
+    g_warning("%s: problem creating km_kbp_state. Status is %u.", __FUNCTION__, status);
+  }
+  g_free(environment_opts);
+  return status;
+}
+
+static km_kbp_status
+load_keyboard_options(IBusKeymanEngine *keyman)
+{
+  g_assert(keyman);
+
+  // Retrieve keyboard options from DConf
+  // TODO: May need unique packageID and keyboard ID
+  g_message("%s: Loading options for kb_name: %s", __FUNCTION__, keyman->kb_name);
+  GQueue *queue_options = keyman_get_options_queue_fromdconf(keyman->kb_name, keyman->kb_name);
+  int num_options       = g_queue_get_length(queue_options);
+  if (num_options < 1)
+    return KM_KBP_STATUS_OK;
+
+  // Allocate enough options for: num_options plus 1 pad struct of 0's
+  km_kbp_option_item *keyboard_opts = g_new0(km_kbp_option_item, num_options + 1);
+
+  for (int i = 0; i < num_options; i++) {
+    km_kbp_option_item *item = g_queue_pop_head(queue_options);
+    keyboard_opts[i].scope = item->scope;
+    keyboard_opts[i].key   = item->key;
+    keyboard_opts[i].value = item->value;
+  }
+
+  keyboard_opts[num_options].scope = 0;
+  keyboard_opts[num_options].key   = 0;
+  keyboard_opts[num_options].value = NULL;
+
+  // once we have the option list we can then update the options using the public api call
+  km_kbp_status status = km_kbp_state_options_update(keyman->state, keyboard_opts);
+
+  if (status != KM_KBP_STATUS_OK) {
+    g_warning("%s: problem creating km_kbp_state. Status is %u.", __FUNCTION__, status);
+  }
+  for (int i = 0; i < num_options; i++) {
+    g_free((km_kbp_cp *)keyboard_opts[i].key);
```

Comment:
Should `scope` also be freed? (I notice the original code doesn't free it either)

Reply:
no, `scope` is a simple `int`, so we don't have to free it. Whereas `key` and `value` are pointers.